2020-01-Winter/2_string_manipulation/README.md

```diff
@@ -116,21 +133,106 @@ Go to [Top](#top)
 <!-- Don't remove -->
 <a name="s2"/>
 
-### 2. SOLUTION 2 TODO :bug:
-
-Source: TODO :bug:
-
-#### Naive/Simple Solution
-
-TODO :bug:
+### 2. GROUP ANAGRAMS
+
+Source:  [LeetCode](https://leetcode.com/problems/group-anagrams/)
+
+Two strings are anagrams if their sorted strings are equal. Then, one way
+to solve this problem is by categorizing by sorting arrays.
+
+1. Check for an empty array, if the array is not empty, then use a hashtable
+where the key is the sorted string, and the value is the string that once is sorted, 
+it matches the key.
+
+2. Iterate through the array of strings. For each string in the array of strings,
+create an array of chars, using the .toCharArray() function. 
+
+3. Sort the char array, and convert it into a string. This string will be used as the
+key for the key-value pair of the hashtable.
+
+4. If the hashtable does not contains the key(string value of sorted char array), add it to the map. 
+
+5. Add the original (unsorted) string as a new value of the key.
+ 
+6. Return the list of list of anagrams
+
+```
+ public static List<List<String>> groupAnagrams(String[] strs) {
+        if (strs.length == 0){
+            return new ArrayList<>();
+        }
+        
+        Map<String, List> map = new HashMap<>();
+        
+        for (String st:strs){
+            
+            char[] chars = st.toCharArray();
+            
+            Arrays.sort(chars);
+            
+            String key = String.valueOf(chars); // Converting the sorted char sequence into string
+            
+            if (!map.containsKey(key)) {
+                map.put(key, new ArrayList());
+            }
+            
+            map.get(key).add(st);
+        }
+        return new ArrayList(map.values());
+    }
+	```
+Time Complexity: The outer loop has complexity O(N) as we iterate through each string. 
+Then, we sort each string in O(K log K) time, where K is the length of each string.
+For many entries, we can assume K is the average length, resulting in a total complexity
+of O(N * K log K). However, K can be treated as a constant if we take it as the average 
+string length in our array. Then, N would be the best (greater) complexity indicator,
+resulting in complexity of O( N log N).
```

Comment:
If `K log K` amortizes to a constant, the resulting complexity should be `O(N)`, no?
```suggestion
resulting in complexity of O( N ).
```